API/projectClass/backgroundsClass.m

```diff
@@ -79,37 +80,45 @@
                     throw(exceptions.invalidNumberOfInputs(sprintf('For type ''%s'', at least three inputs are required, but only %d are supplied', typeVal, length(in))));
                end
 
-               thisRow{1} = in{1};
-               thisRow{2} = in{2};
+               newRow{1} = in{1};
+               newRow{2} = in{2};
 
                % Check that the other params inputted are either valid
-               % background names, or numbers in range..
+               % background names, or numbers in range
                switch typeVal
                    case allowedTypes.Constant.value
-                       % Param 3 must be a valid parameter
-                       thisParam = obj.validateParam(in(3));
-                       thisRow{3} = thisParam;
+                       % Param 3 (source) must be a valid background parameter
+                       newRow{3} = obj.validateParam(in(3));
+
+                   case allowedTypes.Data.value
+                       % Background is contained within a data file.
+                       % We don't have access to the data files at this
+                       % point so this will be checked downstream.
+                       % We also allow for an optional data offset
+                       newRow{3} = in{3};
+                       if length(in) >= 4
+                           newRow{4} = obj.validateParam(in(4));
+                       end
 
                    case allowedTypes.Function.value
-                       throw(exceptions.notImplemented('Function backgrounds are not yet supported.'))
-                       % Param 3 is assumed to be function name
-                       % any other given parameters must be in paramNames
-                       % list or numbers in range
-                       thisRow{3} = in{3};
-                       for i = 4:length(in)
-                          thisParam = obj.validateParam(in(i));
-                          thisRow{i} = thisParam;
+                       % Param 3 (source) is the function name, defined in
+                       % the custom files table
+                       newRow{3} = in{3};
+
+                       if length(in) >= 4
+                           % Any other given parameters must be valid
+                           % background parameters
+                           params = in(4:end);
+                           params = params(~(cellfun(@(x) isequal(x,""), params)));
+                           for i = 1:length(params)
+                              thisParam = obj.validateParam(params(i));
```

Comment:
Hmm, that sounds like a recipe for spaghetti code to me. Because of the unique way the matlab API of RAT is written, we have no idea if the data/custom file entry is even defined when the background is defined. Ultimately, this is why background parameters and backgrounds are defined in the same class. Running this check in `parseClassToStructs` feels the only sensible way to proceed to me.

Reply:
I am just concerned about the inconsistency where we validate params on setting them but don't validate the data or custom files. I don't think holding references to the files and data objects and checking names will lead to spaghetti code except I am missing something. Note background params are not defined in the class they are passed in to the class on initialization https://github.com/RascalSoftware/RAT/blob/fcc187dd175291638c3628cc7a30605fa39f6078/API/projectClass/backgroundsClass.m#L30-L38

Reply:
I don't see the fundamental difference these objects are referred to by the backgrounds so background holds a reference if you are concerned about access to the reference they could be made private. Passing allNames to the add and set function seems like a round about way to do the same thing

Reply:
The point I'm making is that `project.backgrounds.backgroundParams` is defined, but `project.backgrounds.data` isn't.

Reply:
I am not sure how that affect the decision to hold a reference to the 2 objects. The references could be private so no one can access it from the background object

Reply:
Right, I'll look into this further. But the point is that data and custom files cannot be handled identically to background parameters, because they are separate data blocks defined outside of the scope of the background object.